File: elixir/lib/day5.ex
defmodule Day5 do
    def run do
        passes = Input.readFile("day5")
            |> Enum.filter(&(&1 != ""))
            |> Enum.map(&(String.split(&1, "")
            |> Enum.filter(fn(x) -> x != "" end)))
        part1(passes)
        part2(passes)
    end

    defp parseInstruction(instruction, upper, lower, {low, high}) do
        size = high - low
        case instruction do
            i when i == upper -> {low, div(size, 2) + low}
            i when i == lower -> {div(size, 2) + rem(size, 2) + low, high}
            _ -> {low, high}
        end
    end

    defp getBoardingIds(passes) do
        passes
            |> Enum.reduce([],fn(pass, ids) ->
                {row, _} = pass
                    |> Enum.reduce({0, 127}, fn(i, range) ->
                        parseInstruction(i, "F", "B", range)
                    end)
                {col, _} = pass
                    |> Enum.reduce({0, 7}, fn(i, range) ->
                        parseInstruction(i, "L", "R", range)
                    end)
                ids ++ [row * 8 + col]
            end)
    end

    defp part1(passes) do
        high = getBoardingIds(passes) |> Enum.max
        IO.puts "5-1: The highest id is [#{high}]"
    end

    defp part2(passes) do
        ids = getBoardingIds(passes)
        Enum.each(0..127, fn(r) ->
            Enum.each(0..8, fn(c) ->
                id = r * 8 + c
                if not (ids |> Enum.member?(id)) do
                    if ids |> Enum.member?(id - 1) and ids |> Enum.member?(id + 1) do
                        IO.puts "5-2: Your boarding ID is [#{id}]"
                    end
                end
            end)
        end)
    end
end

File: src/day5.rs
use crate::util::read_file;
use fancy_regex::Regex;
use itertools::Itertools;

pub fn run() {
    println!("---- DAY 5 ----");
    part_1();
    part_2();
}

fn get_ids(lines: String) -> Vec<i32> {
    let mut ids = vec![];
    for pass in lines.split('\n') {
        let mut r_low = 0usize;
        let mut r_high = 127usize;
        let mut c_low = 0usize;
        let mut c_high = 7usize;
        for instruction in pass.split("").filter(|x| x != &"").collect::<Vec<&str>>() {
            match instruction {
                "F" => r_high = ((r_high - r_low) / 2) + r_low,
                "B" => r_low = ((r_high - r_low) / 2) + ((r_high - r_low) % 2) + r_low,
                "L" => c_high = ((c_high - c_low) / 2) + c_low,
                "R" => c_low = ((c_high - c_low) / 2) + ((c_high - c_low) % 2) + c_low,
                _ => continue
            };
        }
        assert_eq!(r_low, r_high);
        assert_eq!(c_low, c_high);
        ids.push((r_low * 8 + c_low) as i32);
    }
    ids
}

fn part_1() {
    let lines = read_file("inputs/day5.txt");
    println!("Highest ID is [{:?}]", get_ids(lines).into_iter().max().unwrap());
}

fn part_2() {
    let lines = read_file("inputs/day5.txt");
    let ids = get_ids(lines);
    for r in 0i32..127 {
        for c in 0i32..7 {
            let id = r * 8 + c;
            if !ids.contains(&id) && ids.contains(&(id - 1)) && ids.contains(&(id + 1)) {
                println!("Your boarding ID is [{}]", id);
            }
        }
    }
}